plato/trainers/basic.py

```diff
@@ -447,3 +509,15 @@ async def server_test(self, testset, sampler=None):
                 await asyncio.sleep(0)
 
         return correct / total
+
+    def obtain_model_update(self, wall_time):
+        """
+            Obtain a saved model for a particular epoch that finishes just after the provided
+            wall clock time is reached.
+        """
+        print("len of models per epoch = ", len(self.models_per_epoch))
+        print(self.models_per_epoch)
+        for epoch in sorted(self.models_per_epoch):
+            model = self.models_per_epoch[epoch]
+            if model['time'] + self.tic > wall_time:
+                return model['model']
```

Comment:
Just some questions about the urgent message implementations. Why do we need to find the model per epoch by the wall clock time? I think here we want to find the model that is trained by the most recent epoch, so could we just return the model with the largest epoch number in `self.models_per_epoch`?

Thanks!

Reply:
The wall clock time was previously provided by the server, and that was the simulated wall clock time when the previous client (before the one that the urgent request needs to be sent to) had finished training. This is when the server needs to traverse all the training clients and send them the urgent request to stop training prematurely.

However, the current implementation does have a bug: it assumes that `train_process()` is sharing memory with its parent process, which is not the case. You can see from debugging output that the length of self.models_per_epoch is zero. This is incorrect.

To fix this, we will (unfortunately) have to "remember" these models by writing to checkpoints, and then reading from one these checkpoints when `obtain_model_update()` is called.

Reply:
I see. That makes sense. Thanks!